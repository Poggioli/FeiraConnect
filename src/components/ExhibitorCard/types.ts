import { GetExhibitorsByStreetMarket } from "@/services/getExhibitorsByStreetMarket";

export type ExhibitorCardProps = GetExhibitorsByStreetMarket;

export type UseExhibitorCardPropsProps = Pick<ExhibitorCardProps, "name" | "phone" | "whatsApp" | "website">;
