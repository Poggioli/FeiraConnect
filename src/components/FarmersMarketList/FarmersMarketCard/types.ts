import { Frequency, SearchFarmersMarketByCityId } from "@/services/searchFarmersMarketByCityId";

export type FarmersMarketCardProps = SearchFarmersMarketByCityId

export type UseFarmersMarketCardProps = {
  frequency: Frequency,
  startTime: Date,
  endTime: Date
}