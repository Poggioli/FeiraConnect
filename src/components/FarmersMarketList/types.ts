export type FarmersMarketListProps = {
  city: string;
};

export type UseFarmersMarketListProps = {
  city: string
}