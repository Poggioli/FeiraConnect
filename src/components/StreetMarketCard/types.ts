import { GetStreetMarketsByCity } from "@/services/getStreetMarketsByCity";

export type StreetMarketCardProps = GetStreetMarketsByCity;

export type UseStreetMarketCardProps = Pick<
  StreetMarketCardProps,
  "name" | "neighborhood" | "weekday" | "apperture" | "closure"
>;
