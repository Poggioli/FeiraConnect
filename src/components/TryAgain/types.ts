export type TryAgainProps = {
  refetch: () => unknown;
}