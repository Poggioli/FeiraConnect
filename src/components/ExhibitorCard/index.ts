export * from "./ExhibitorCard";
export * from "./useExhibitorCard";
export * from "./types";