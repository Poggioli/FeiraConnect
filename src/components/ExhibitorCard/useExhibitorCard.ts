import { UseExhibitorCardPropsProps } from "./types";
import parsePhoneNumber, { PhoneNumber } from "libphonenumber-js";

export function useExhibitorCard({
  name,
  phone,
  whatsApp,
  website,
}: UseExhibitorCardPropsProps) {
  const formatedName = name.charAt(0).toUpperCase() + name.slice(1);
  const hasWebsite = !!website;

  const phoneNumber = parsePhoneNumber(phone, "BR") as PhoneNumber;
  const telephone = {
    formatedNumber: `${phoneNumber.formatNational()}`,
    link: `tel:${phoneNumber.formatInternational()}`,
    "aria-label": `Número de telefone de ${formatedName}`,
  };

  const hasWhatsApp = !!whatsApp;
  const wpp = parsePhoneNumber(whatsApp || "", "BR");
  const whatsAppLink = encodeURI(
    `https://wa.me/${wpp?.countryCallingCode}${wpp?.nationalNumber}?text=Olá ${formatedName}, encontrei seu número através do feira-connect.`
  );
  const whatsAppNumber = {
    formatedNumber: `${wpp?.formatNational()}`,
    link: whatsAppLink,
    "aria-label": `Enviar mensagem pelo whatsApp para ${formatedName}`,
  };

  return {
    formatedName,
    telephone,
    whatsAppNumber,
    hasWhatsApp,
    hasWebsite,
  };
}
