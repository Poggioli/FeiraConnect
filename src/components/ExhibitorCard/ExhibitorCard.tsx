import WhatsApp from "@/assets/whatsapp-icon.svg?react";
import { Badge } from "@/components/ui/badge";
import { Card, CardDescription, CardTitle } from "@/components/ui/card";
import { cn } from "@/lib/utils";
import { Earth, Phone } from "lucide-react";
import { FC, HTMLAttributes, forwardRef } from "react";
import { ExhibitorCardProps } from "./types";
import { useExhibitorCard } from "./useExhibitorCard";
import { Skeleton } from "@/components/ui/skeleton";

export const ExhibitorCard = forwardRef<
  HTMLDivElement,
  HTMLAttributes<HTMLDivElement> & ExhibitorCardProps
>(({ name, description, phone, whatsApp, website, items, ...props }, ref) => {
  const {
    formatedName,
    telephone,
    whatsAppNumber,
    hasWhatsApp
  } = useExhibitorCard({ name, phone, whatsApp });

  return (
    <Card {...props} className={cn("py-4 px-5 w-full flex flex-col", props.className)} ref={ref}>
      <CardTitle>{formatedName}</CardTitle>
      {description ? <CardDescription>{description}</CardDescription> : null}
      <div className="flex flex-col justify-between flex-1">
        <div className="mt-2 flex flex-row justify-start items-center flex-wrap gap-1">
          {items.map((item) => (
            <Badge variant="outline" key={item}>
              {item}
            </Badge>
          ))}
        </div>
        <div className="mt-6 flex flex-row justify-end items-start flex-wrap gap-3">
          <a
            href={telephone.link}
            target="_blank"
            className="leading-[initial]"
            aria-label={telephone["aria-label"]}
          >
            <Badge variant="secondary" className="gap-2">
              <Phone className="h-3 w-3" />
              {telephone.formatedNumber}
            </Badge>
          </a>
          {hasWhatsApp ? (
            <a
              href={whatsAppNumber.link}
              target="_blank"
              className="leading-[initial]"
              aria-label={whatsAppNumber["aria-label"]}
            >
              <Badge variant="secondary" className="gap-2">
                <WhatsApp className="h-3 w-3 fill-current" />
                {whatsAppNumber.formatedNumber}
              </Badge>
            </a>
          ) : null}
          {website ? (
            <a href={website} target="_blank" className="leading-[initial]">
              <Badge variant="secondary" className="gap-2">
                <Earth className="h-3 w-3" />
                Site
              </Badge>
            </a>
          ) : null}
        </div>
      </div>
    </Card>
  );
});

export const ExhibitorCardSkeleton: FC = () => (
  <Card className="py-4 px-5 w-full flex flex-col">
    <div className="flex flex-col justify-center items-start gap-1">
      <Skeleton className="rounded-sm h-4 w-36" />
      <Skeleton className="rounded-sm h-[0.875rem] w-full" />
    </div>
    <div className="flex flex-col justify-between flex-1">
      <div className="mt-2 flex flex-row justify-start items-center flex-wrap gap-1">
        {Array.from(Array(5).keys()).map((i) => <Skeleton key={i} className="rounded-sm h-4 w-24" />)}
      </div>
      <div className="mt-6 flex flex-row justify-end items-start flex-wrap gap-3">
        <Skeleton className="rounded-sm h-4 w-36" />
        <Skeleton className="rounded-sm h-4 w-36" />
        <Skeleton className="rounded-sm h-4 w-16" />
      </div>
    </div>
  </Card>
)