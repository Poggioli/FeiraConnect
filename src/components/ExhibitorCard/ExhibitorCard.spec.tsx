import { render } from "@testing-library/react";
import { ExhibitorCard, ExhibitorCardSkeleton } from "./ExhibitorCard";
import { ExhibitorCardProps } from "./types";

describe("@/components/ExhibitorCard", () => {
  const defaultProps: ExhibitorCardProps = {
    id: "id-exhibitor-name",
    slug: "exhibitor-name",
    items: ["item 1", "item 2", "item 3"],
    description: "description",
    name: "exhibitor name",
    phone: "11912345678",
    whatsApp: "11912345679",
    website: "https://exhibitor-name.com.br",
  };

  it(`GIVEN a ExhibitorCard
      WHEN render
      THEN should show the formatedName`, () => {
    const { getByText } = render(<ExhibitorCard {...defaultProps} />);
    const formatedName = getByText("Exhibitor name");
    expect(formatedName).toBeVisible();
  });

  describe("description", () => {
    it(`GIVEN a ExhibitorCard
        WHEN description is truthy value
        THEN should show the description value`, () => {
      const { getByText } = render(<ExhibitorCard {...defaultProps} />);
      const description = getByText("description");
      expect(description).toBeVisible();
    });

    it(`GIVEN a ExhibitorCard
        WHEN description is falsy value
        THEN should NOT show the description value`, () => {
      const { baseElement } = render(<ExhibitorCard {...defaultProps} description={undefined} />);
      expect(baseElement).toMatchSnapshot();
    });
  });

  it(`GIVEN a ExhibitorCard
      WHEN render
      THEN should show the items value`, () => {
    const { getByText } = render(<ExhibitorCard {...defaultProps} />);
    const item1 = getByText("item 1");
    const item2 = getByText("item 2");
    const item3 = getByText("item 3");
    expect(item1).toBeVisible();
    expect(item2).toBeVisible();
    expect(item3).toBeVisible();
  });

  describe("telephone", () => {
    it(`GIVEN a ExhibitorCard
        WHEN render
        THEN should show the anchor tag with href to tel`, () => {
      const { getByLabelText } = render(<ExhibitorCard {...defaultProps} />);
      const anchorTag = getByLabelText("Número de telefone de Exhibitor name");
      expect(anchorTag).toHaveAttribute("href", "tel:+55 11 91234 5678");
    });

    it(`GIVEN a ExhibitorCard
        WHEN render
        THEN should show the anchor tag with aria-label`, () => {
      const { getByLabelText } = render(<ExhibitorCard {...defaultProps} />);
      const ariaLabel = getByLabelText("Número de telefone de Exhibitor name");
      expect(ariaLabel).toBeVisible();
    });

    it(`GIVEN a ExhibitorCard
        WHEN render
        THEN should show the formated number`, () => {
      const { getByText } = render(<ExhibitorCard {...defaultProps} />);
      const formatedNumber = getByText("(11) 91234-5678");
      expect(formatedNumber).toBeVisible();
    });
  });

  describe("whatsAppNumber", () => {
    it(`GIVEN a ExhibitorCard
        WHEN render
        THEN should show the anchor tag with href to tel`, () => {
      const { getByLabelText } = render(<ExhibitorCard {...defaultProps} />);
      const anchorTag = getByLabelText("Enviar mensagem pelo whatsApp para Exhibitor name");
      expect(anchorTag).toHaveAttribute("href", "https://wa.me/5511912345679?text=Ol%C3%A1%20Exhibitor%20name,%20encontrei%20seu%20n%C3%BAmero%20atrav%C3%A9s%20do%20feira-connect.");
    });

    it(`GIVEN a ExhibitorCard
        WHEN render
        THEN should show the anchor tag with aria-label`, () => {
      const { getByLabelText } = render(<ExhibitorCard {...defaultProps} />);
      const ariaLabel = getByLabelText("Enviar mensagem pelo whatsApp para Exhibitor name");
      expect(ariaLabel).toBeVisible();
    });

    it(`GIVEN a ExhibitorCard
        WHEN render
        THEN should show the formated number`, () => {
      const { getByText } = render(<ExhibitorCard {...defaultProps} />);
      const formatedNumber = getByText("(11) 91234-5679");
      expect(formatedNumber).toBeVisible();
    });

    it(`GIVEN a ExhibitorCard
        WHEN whatsApp number is falsy
        THEN should NOT show the whatsApp tag`, () => {
      const { queryByText } = render(<ExhibitorCard {...defaultProps} whatsApp={undefined} />);
      const formatedNumber = queryByText("(11) 91234-5679");
      expect(formatedNumber).not.toBeInTheDocument();
    });
  });

  describe("website", () => {
    it(`GIVEN a ExhibitorCard
        WHEN render
        THEN should show the site tag`, () => {
      const { getByText } = render(<ExhibitorCard {...defaultProps} />);
      const tag = getByText("Site");
      expect(tag).toBeVisible();
    });

    it(`GIVEN a ExhibitorCard
        WHEN website is falsy
        THEN should NOT show the site tag`, () => {
      const { queryByText } = render(<ExhibitorCard {...defaultProps} website={undefined} />);
      const tag = queryByText("Site");
      expect(tag).not.toBeInTheDocument();
    });
  });

  it(`GIVEN a ExhibitorCardSkeleton
      WHEN render
      then should everything rigth`, () => {
    const { baseElement } = render(<ExhibitorCardSkeleton />);
    expect(baseElement).toMatchSnapshot();
  })
});