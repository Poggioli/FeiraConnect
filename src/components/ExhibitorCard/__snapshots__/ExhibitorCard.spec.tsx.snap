// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@/components/ExhibitorCard > GIVEN a ExhibitorCardSkeleton
      WHEN render
      then should everything rigth 1`] = `
<body>
  <div>
    <div
      class="rounded-xl border bg-card text-card-foreground py-4 px-5 w-full flex flex-col"
    >
      <div
        class="flex flex-col justify-center items-start gap-1"
      >
        <div
          class="animate-pulse bg-primary/10 rounded-sm h-4 w-36"
        />
        <div
          class="animate-pulse bg-primary/10 rounded-sm h-[0.875rem] w-full"
        />
      </div>
      <div
        class="flex flex-col justify-between flex-1"
      >
        <div
          class="mt-2 flex flex-row justify-start items-center flex-wrap gap-1"
        >
          <div
            class="animate-pulse bg-primary/10 rounded-sm h-4 w-24"
          />
          <div
            class="animate-pulse bg-primary/10 rounded-sm h-4 w-24"
          />
          <div
            class="animate-pulse bg-primary/10 rounded-sm h-4 w-24"
          />
          <div
            class="animate-pulse bg-primary/10 rounded-sm h-4 w-24"
          />
          <div
            class="animate-pulse bg-primary/10 rounded-sm h-4 w-24"
          />
        </div>
        <div
          class="mt-6 flex flex-row justify-end items-start flex-wrap gap-3"
        >
          <div
            class="animate-pulse bg-primary/10 rounded-sm h-4 w-36"
          />
          <div
            class="animate-pulse bg-primary/10 rounded-sm h-4 w-36"
          />
          <div
            class="animate-pulse bg-primary/10 rounded-sm h-4 w-16"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`@/components/ExhibitorCard > description > GIVEN a ExhibitorCard
        WHEN description is falsy value
        THEN should NOT show the description value 1`] = `
<body>
  <div>
    <div
      class="rounded-xl border bg-card text-card-foreground py-4 px-5 w-full flex flex-col"
      id="id-exhibitor-name"
      slug="exhibitor-name"
    >
      <h3
        class="font-semibold leading-none tracking-tight"
      >
        Exhibitor name
      </h3>
      <div
        class="flex flex-col justify-between flex-1"
      >
        <div
          class="mt-2 flex flex-row justify-start items-center flex-wrap gap-1"
        >
          <div
            class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-foreground"
          >
            item 1
          </div>
          <div
            class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-foreground"
          >
            item 2
          </div>
          <div
            class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-foreground"
          >
            item 3
          </div>
        </div>
        <div
          class="mt-6 flex flex-row justify-end items-start flex-wrap gap-3"
        >
          <a
            aria-label="NÃºmero de telefone de Exhibitor name"
            class="leading-[initial]"
            href="tel:+55 11 91234 5678"
            target="_blank"
          >
            <div
              class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80 gap-2"
            >
              <svg
                class="lucide lucide-phone h-3 w-3"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M22 16.92v3a2 2 0 0 1-2.18 2 19.79 19.79 0 0 1-8.63-3.07 19.5 19.5 0 0 1-6-6 19.79 19.79 0 0 1-3.07-8.67A2 2 0 0 1 4.11 2h3a2 2 0 0 1 2 1.72 12.84 12.84 0 0 0 .7 2.81 2 2 0 0 1-.45 2.11L8.09 9.91a16 16 0 0 0 6 6l1.27-1.27a2 2 0 0 1 2.11-.45 12.84 12.84 0 0 0 2.81.7A2 2 0 0 1 22 16.92z"
                />
              </svg>
              (11) 91234-5678
            </div>
          </a>
          <a
            aria-label="Enviar mensagem pelo whatsApp para Exhibitor name"
            class="leading-[initial]"
            href="https://wa.me/5511912345679?text=Ol%C3%A1%20Exhibitor%20name,%20encontrei%20seu%20n%C3%BAmero%20atrav%C3%A9s%20do%20feira-connect."
            target="_blank"
          >
            <div
              class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80 gap-2"
            >
              <svg
                class="h-3 w-3 fill-current"
                height="800px"
                id="Layer_1"
                viewBox="0 0 308 308"
                width="800px"
                xml:space="preserve"
                xmlns="http://www.w3.org/2000/svg"
                xmlns:xlink="http://www.w3.org/1999/xlink"
              >
                <g
                  id="XMLID_468_"
                >
                  <path
                    d="M227.904,176.981c-0.6-0.288-23.054-11.345-27.044-12.781c-1.629-0.585-3.374-1.156-5.23-1.156 c-3.032,0-5.579,1.511-7.563,4.479c-2.243,3.334-9.033,11.271-11.131,13.642c-0.274,0.313-0.648,0.687-0.872,0.687 c-0.201,0-3.676-1.431-4.728-1.888c-24.087-10.463-42.37-35.624-44.877-39.867c-0.358-0.61-0.373-0.887-0.376-0.887 c0.088-0.323,0.898-1.135,1.316-1.554c1.223-1.21,2.548-2.805,3.83-4.348c0.607-0.731,1.215-1.463,1.812-2.153 c1.86-2.164,2.688-3.844,3.648-5.79l0.503-1.011c2.344-4.657,0.342-8.587-0.305-9.856c-0.531-1.062-10.012-23.944-11.02-26.348 c-2.424-5.801-5.627-8.502-10.078-8.502c-0.413,0,0,0-1.732,0.073c-2.109,0.089-13.594,1.601-18.672,4.802 c-5.385,3.395-14.495,14.217-14.495,33.249c0,17.129,10.87,33.302,15.537,39.453c0.116,0.155,0.329,0.47,0.638,0.922 c17.873,26.102,40.154,45.446,62.741,54.469c21.745,8.686,32.042,9.69,37.896,9.69c0.001,0,0.001,0,0.001,0 c2.46,0,4.429-0.193,6.166-0.364l1.102-0.105c7.512-0.666,24.02-9.22,27.775-19.655c2.958-8.219,3.738-17.199,1.77-20.458 C233.168,179.508,230.845,178.393,227.904,176.981z"
                    id="XMLID_469_"
                  />
                  <path
                    d="M156.734,0C73.318,0,5.454,67.354,5.454,150.143c0,26.777,7.166,52.988,20.741,75.928L0.212,302.716 c-0.484,1.429-0.124,3.009,0.933,4.085C1.908,307.58,2.943,308,4,308c0.405,0,0.813-0.061,1.211-0.188l79.92-25.396 c21.87,11.685,46.588,17.853,71.604,17.853C240.143,300.27,308,232.923,308,150.143C308,67.354,240.143,0,156.734,0z  M156.734,268.994c-23.539,0-46.338-6.797-65.936-19.657c-0.659-0.433-1.424-0.655-2.194-0.655c-0.407,0-0.815,0.062-1.212,0.188 l-40.035,12.726l12.924-38.129c0.418-1.234,0.209-2.595-0.561-3.647c-14.924-20.392-22.813-44.485-22.813-69.677 c0-65.543,53.754-118.867,119.826-118.867c66.064,0,119.812,53.324,119.812,118.867 C276.546,215.678,222.799,268.994,156.734,268.994z"
                    id="XMLID_470_"
                  />
                </g>
              </svg>
              (11) 91234-5679
            </div>
          </a>
          <a
            class="leading-[initial]"
            href="https://exhibitor-name.com.br"
            target="_blank"
          >
            <div
              class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80 gap-2"
            >
              <svg
                class="lucide lucide-earth h-3 w-3"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M21.54 15H17a2 2 0 0 0-2 2v4.54"
                />
                <path
                  d="M7 3.34V5a3 3 0 0 0 3 3v0a2 2 0 0 1 2 2v0c0 1.1.9 2 2 2v0a2 2 0 0 0 2-2v0c0-1.1.9-2 2-2h3.17"
                />
                <path
                  d="M11 21.95V18a2 2 0 0 0-2-2v0a2 2 0 0 1-2-2v-1a2 2 0 0 0-2-2H2.05"
                />
                <circle
                  cx="12"
                  cy="12"
                  r="10"
                />
              </svg>
              Site
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</body>
`;
